types/src/params.rs

```diff
@@ -408,6 +416,275 @@ impl<'a> Id<'a> {
 	}
 }
 
+/// Helper module for building parameters.
+mod params_builder {
+	use serde::Serialize;
+
+	/// Initial number of bytes for a parameter length.
+	const PARAM_BYTES_CAPACITY: usize = 128;
+
+	/// Generic parameter builder that serializes parameters to bytes.
+	/// This produces a JSON compatible String.
+	///
+	/// The implementation relies on `Vec<u8>` to hold the serialized
+	/// parameters in memory for the following reasons:
+	///   1. Other serialization methods than `serde_json::to_writer` would internally
+	///      have an extra heap allocation for temporarily holding the value in memory.
+	///   2. `io::Write` is not implemented for `String` required for serialization.
+	#[derive(Debug)]
+	pub struct ParamsBuilder {
+		bytes: Vec<u8>,
+		end: char,
+	}
+
+	impl ParamsBuilder {
+		/// Construct a new [`ParamsBuilder`] with custom start and end tokens.
+		/// The inserted values are wrapped by the _start_ and _end_ tokens.
+		fn new(start: char, end: char) -> Self {
+			let mut bytes = Vec::with_capacity(PARAM_BYTES_CAPACITY);
+			bytes.push(start as u8);
+			ParamsBuilder { bytes, end }
+		}
+
+		/// Construct a new [`ParamsBuilder`] for positional parameters equivalent to a JSON array object.
+		pub fn positional() -> Self {
+			Self::new('[', ']')
+		}
+
+		/// Construct a new [`ParamsBuilder`] for named parameters equivalent to a JSON map object.
+		pub fn named() -> Self {
+			Self::new('{', '}')
+		}
+
+		/// Insert a named value (key, value) pair into the builder.
+		/// The _name_ and _value_ are delimited by the `:` token.
+		pub fn insert_named<P: Serialize>(&mut self, name: &str, value: P) -> Result<(), serde_json::Error> {
+			serde_json::to_writer(&mut self.bytes, name)?;
+			self.bytes.push(b':');
+			serde_json::to_writer(&mut self.bytes, &value)?;
+			self.bytes.push(b',');
+
+			Ok(())
+		}
+
+		/// Insert a plain value into the builder.
+		pub fn insert<P: Serialize>(&mut self, value: P) -> Result<(), serde_json::Error> {
+			serde_json::to_writer(&mut self.bytes, &value)?;
+			self.bytes.push(b',');
+
+			Ok(())
+		}
+
+		/// Finish the building process and return a JSON compatible string.
+		pub fn build(mut self) -> String {
```

Comment:
Nit: we allocate even if no params are provided (but the ToRpcParams trait allows us to avoid that by returning an Option).

Perhaps this should return `Option<String>`, and avoid allocating itself if no params are given, to take advantage of that?

Reply:
I was thinking that general users will want to use the `rpc_params` macro for building the parameters, as opposed to opting for the builders themselves.
While at the same time, experience users might want to implement `ToRpcParams` for their needs.

A nice thing about it is that it can generate either: `EmptyParams` (no allocation if no arguments are provided), `Unnamed/Named Params` (which does allocate). This way the `Unnamed/Named` are committed to allocation, since we expect at least one parameter and hence the `128` bytes allocation to avoid resizing. 

Do you mean to modify it in the `ParamsBuilder` or everywhere using it (Named and Unnamed)? 🤔